### ISSUE TITLE
VK_ERROR_DEVICE_LOST on Intel+Wayland

### ISSUE THREAD

**Issue by jekstrand** [hrydgard/ppsspp#13009]
### What happens?

I try to start PPSSPP with the Vulkan back-end on Intel with GNOME+Wayland.  It immediately gets a `VK_ERROR_DEVICE_LOST`.  Sometimes it doesn't print the error and just hangs.  If I can get past that point (running in GDB helps for some unknown reason, things seem to mostly work.

With a debug build of the driver, the type of error is that the kernel is rejecting the command buffer.  The only thing I can think of which would reproduce that error that reliably (I'm a mesa dev) is if you're submitting a command buffer using a memory object, fence, or semaphore which has been deleted.  Given that it happens right at start-up, I'm guessing something is wrong with your `VK_ERROR_OUT_OF_DATE` handling in your window-system code.

### What *should* happen?

It should start up just fine.

### What hardware, operating system, and PPSSPP version?  On desktop, GPU matters for graphical issues.

OS: Linux, Fedora 32
Hardware: Intel HD graphics 5500

**Comment by unknownbrackets** [comment]
Thanks for reporting.

It's certainly possible we're not handling VK_ERROR_OUT_OF_DATE well/properly.  Unfortunate that the validation layer is clearly not catching this...

Since it seems to happen inconsistently, I just want to confirm if disabling the render thread changes this.  It's not something we'd actually want to do, just a diagnosis step.  The code is here:

https://github.com/hrydgard/ppsspp/blob/348213f68989fb8a165aabae660d64a1a9a6400c/ext/native/thin3d/VulkanRenderManager.cpp#L146-L149

Just move `useThread_ = false;` up one line.

If that helps, maybe AMD was also hitting the same bug which means maybe we can solve both these issues at once (maybe I'm being too hopeful.)

Another thing to try would be splitting the submit of the init commands and the draw commands, which can be done by changing `false` to `true` here:
https://github.com/hrydgard/ppsspp/blob/141ad162a244dae525a2a580fcbcc2fc6451ace8/SDL/SDLVulkanGraphicsContext.cpp#L100

This is another thing we wouldn't actually want to do, but that'd help confirm if it dies on the [second submit](https://github.com/hrydgard/ppsspp/blob/348213f68989fb8a165aabae660d64a1a9a6400c/ext/native/thin3d/VulkanRenderManager.cpp#L1103) which is where we do the fence/semaphore stuff.

-[Unknown]

**Comment by hrydgard** [comment]
If we were really doing a mistake like what you suggests with the synchronization objects, the validation layers would yell at us. I'm calling driver issue...

**Comment by jekstrand** [comment]
I think you trust the validation layers far more than I do. :smile: That said, I have a theory about the kind of driver bug this might be (if it is one).  When you get a `VK_ERROR_OUT_OF_DATE_KHR`, do you use the `VkSwapchainCreateInfoKHR::oldSwapchain` parameter to properly chain things together when you create the new one?  And do you have any outstanding acquired images which you then present?  I could imagine us getting the hand-off wrong.

**Comment by hrydgard** [comment]
The validation layers have been immensely helpful in a different effort recently at least! So much better than they used to be, and stuff like this they tend to get right.

That said, we do not currently use oldSwapchain even if we should. Our API teardown/bringup stuff is designed patterned after GLES 2.0 where we have to support tearing everything down to the ground and back up at any time (even though that hasn't been necessary on Android for a long time, we still support Gingerbread ;) 

So maybe I should indeed look into the oldSwapchain stuff.

**Comment by jekstrand** [comment]
If you're tearing down completely, that seems to hint away from driver bugs.  I could see us getting `oldSwapchain` wrong but I don't see how we'd accidentally have an invalid BO or syncobj handle otherwise.

Reproducing this is very easy at least on GNOME shell.  It sometimes happens on start-up and then always happens the moment you resize or alt-tab to another window.

**Comment by hrydgard** [comment]
Well we do make some effort to not run the teardown on window resize on some systems. On Android we do it anyway but window resizes there are mostly device rotations. 

Anyway, does sound like a swapchain issue, and I should probably go ahead and make it use oldSwapchain just to see if that makes a difference. Running a little low on time before Swedish Midsummer though so might be next week.

**Comment by jekstrand** [comment]
No worries.  For now "run in gdb" combined with "don't alt+tab" are a good enough workaround that I'm enjoying my game. :sweat_smile:

**Comment by hrydgard** [comment]
So after looking around a bit, I just realized that on resize, we don't tear down everything but we do, for no good reason, tear down and recreate the surface. That's probably a mistake, although it kinda should work.

I'm going to see how hard it is to get rid of that.

**Comment by hrydgard** [comment]
#13041 refactors that a bit. Hopefully it'll help.

**Comment by jekstrand** [comment]
I did a bit of poking over the week-end but I wasn't able to see what was happening driver-side.  It looks like my theories about early object destruction are wrong but it's hard to tell.  I'll have to run a custom kernel build with additional debugging to figure this one out. :frowning: I'll try to get more concrete data eventually.

**Comment by hrydgard** [comment]
Weird. Looking forward to finding out if we are doing anything wrong :)

**Comment by hrydgard** [comment]
Just recalled another little oddity we do. We do a little Vulkan availability check on startup, actually creating and destroying an instance and a device. Then, we launch Vulkan for real a little later during startup. 

That first instantiation (VulkanMayBeAvailable in Common/Vulkan/VulkanLoader.cpp) might tickle some driver bug.

**Comment by hrydgard** [comment]
I keep getting device losses when alt-tabbing away during fullscreen. Seems surfaces can get into a state where rendering to them crashes inside the driver. In this case, it seems the driver erroneously returns VK_SUBOPTIMAL_KHR instead of VK_ERROR_OUT_OF_DATE.

Here's another app that's affected: https://forum.zdoom.org/viewtopic.php?t=67857&p=1142027

**Comment by jekstrand** [comment]
Built a kernel with extra debug stuff and I figured out what's going wrong.  Someone is waiting on a `VkSemaphore` that has never been signalled.  I hacked up the driver to start all semaphores off signaled and the bug went away.  I see a black flash when I would have seen a crash before.  Best guess is that `vkQueuePresent` is being called without actually rendering to the image.

It's possible that we're failing to signal the semaphore somewhere but there are only two places we would do that and I'm pretty sure we've got it covered both places.

**Comment by hrydgard** [comment]
That is good information, thanks! Very plausible area too since semaphores are both a weak point of the validation layers and in my understanding of Vulkan heh. 

I now recall we have some possibly misguided code that skips presents in some edge cases, it might let a semaphore loose...

**Comment by hrydgard** [comment]
OK so we actually do seem to handle it properly, as far as I can tell. Could you get a stack trace of the bad wait?

Here are my theories:

* When the driver's vkAcquireNextImageKHR fails with VK_SUBOPTIMAL_KHR, it forgets to signal the semaphore. We handle that as a success (which it is, although a suboptimal one), and wait for the semaphore. Although a lot of other apps would also have run into this, you'd think....

* vkAcquireNextImageKHR returned something other than VK_SUCCESS, VK_SUBOPTIMAL_KHR and VK_ERROR_OUT_OF_DATE_KHR. This case we handle as VK_SUCCESS with an assert that it's VK_SUCCESS, but if that assert is compiled out somehow, we could end up in a condition where we wait on an unsignalled semaphore. But the type of assert I thought we compiled in, even in release builds. Hm.

* vkQueueSubmit fails. Similar to the above, we have an assert that should catch this.

I will improve the error handling and logging a bit, and we'll see if something shakes out I guess. I still suspect the driver is not handling VK_SUBOPTIMAL_KHR correctly.

**Comment by hrydgard** [comment]
On another note, after a restart, I stopped getting the crashes I had alt+tabbing from fullscreen mode. I don't know if the driver got auto-updated or something..

**Comment by hrydgard** [comment]
Okay my Vulkan Linux mess is getting worse and worse. Turns out that the fullscreen bug appeared to be gone because my laptop had switched to using the NVIDIA GPU, and the device selection code wasn't working on SDL. Now that I fixed that, when trying to use the Intel device I'm getting the following with an instant process exit that I can't seem to catch:

```
VulkanRenderManager.cpp:219: Starting Vulkan submission thread (threadInitFrame_ = 0)
X Error of failed request:  BadDrawable (invalid Pixmap or Window parameter)
  Major opcode of failed request:  149 ()
  Minor opcode of failed request:  4
  Resource id in failed request:  0x220000c
  Serial number of failed request:  225
  Current serial number in output stream:  235

```

Really not sure how I'm supposed to be able to work with this... :(

**Comment by hrydgard** [comment]
Just found out that SDL has its own Vulkan initialization functions (https://wiki.libsdl.org/CategoryVulkan), I've just used the X window handle and done the usual stuff on my own. But that should also work.

**Comment by jekstrand** [comment]
> When the driver's vkAcquireNextImageKHR fails with VK_SUBOPTIMAL_KHR, it forgets to signal the semaphore. We handle that as a success (which it is, although a suboptimal one), and wait for the semaphore. Although a lot of other apps would also have run into this, you'd think....

Turns out this was spot-on.  I think the reason why we've not seen it in the past is because we really only return suboptimal on Wayland and there are very few non-trivial Vulkan apps which support Wayland natively.  There are some demos of course but they may not be doing complex enough things to hit this.  In any case, I've made a Mesa MR so this bug can be closed:

https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/5672

Turns out this bug has been around in one form or another since I very first wrote the Vulkan WSI code back in 2015.... :frowning: I think it's only been since about december that this latent bug has had such dire consequences.

**Comment by jekstrand** [comment]
As far as your hybrid graphics woes go... Nvidia+Intel works fine with X11 and the X server running on the Intel card.  That's my default setup these days.  I've had trouble with Wayland working on this laptop thanks to the nvidia card (I've been using PPSSPP on an older intel-only laptop).  I'm not sure about what happens if you have the X server or Wayland compositor running on the Nvidia card.  I think it should work with nouveau but I wouldn't count on it with the blob driver.

**Comment by hrydgard** [comment]
Another one bites the dust! Thanks for fixing it, hope it also fixes my coming-out-of-fullscreen issue. Turning off the NVIDIA GPU in nvidia-settings made it return (of course).

I'm on Ubuntu 20.04 like I said, no idea if I'm running X or wayland but given the messages probably X? It would be nice if the Intel device worked in the default setup (which, after a couple of reboots, turned out to be NVIDIA on this Asus Zenbook). Those X errors that I pasted above are nasty because the process exits with me entirely unable to catch it, so I can't automatically switch back to the NVIDIA GPU if it happens :(

**Comment by hrydgard** [comment]
That's by the way a 2 out of 2, when it comes to issues you have opened in PPSSPP that turned out to be the driver ;)

**Comment by jekstrand** [comment]
> That's by the way a 2 out of 2, when it comes to issues you have opened in PPSSPP that turned out to be the driver ;) But keep em coming, stuff gets fixed either way.

Congratulations.  You're doing much better than average.  I'd say about 60% of all bugs filed against the Intel Linux Vulkan driver actually turn out to be app bugs.  This is especially true of synchronization issues.

> I'm on Ubuntu 20.04 like I said, no idea if I'm running X or wayland but given the messages probably X? It would be nice if the Intel device worked in the default setup (which, after a couple of reboots, turned out to be NVIDIA on this Asus Zenbook). Those X errors that I pasted above are nasty because the process exits with me entirely unable to catch it, so I can't automatically switch back to the NVIDIA GPU if it happens :(

Yes, sadly, some Intel+Nvidia laptops wire the displays to the Nvidia card.  The Intel card can still render fine but Nvidia doesn't support the right X11 extensions required for Intel to display there.  They do support them from the other side so you can render on Nvidia and display on Intel but going the other way doesn't work.  There's very little that I or anyone not at Nvidia can do about that, sadly. :frowning:

**Comment by hrydgard** [comment]
So can I programmatically find out about that? Or actually, I'm not sure our checks that we get a queue that can present are robust enough. Is that what happens then, that I don't get one?

**Comment by hrydgard** [comment]
Do you by the way happen to know who can I file a bug against to get this kind of error to return an error code somewhere instead of killing the process?

**Comment by jekstrand** [comment]
> So can I programmatically find out about that? Or actually, I'm not sure our checks that we get a queue that can present are robust enough. Is that what happens then, that I don't get one? I guess I'll go check..

In theory, you should be able to call `vkGetPhysicalDeviceSurfaceSupportKHR()` and it should return false if trying to display on Nvidia.  I worked on that a long time ago and, at least at the time, the user I had testing it for me said it worked.  I've never had a setup where I can actually test it though.

Still on the topic of swapchains, we're still not quite out of the woods yet, I'm afraid.  If I start up PPSSPP without running it in GDB, I get `VK_ERROR_OUT_OF_DATE` immediately and this doesn't seem to be handled properly.  The app doesn't crash but it also doesn't resize and re-create the swapchain as needed.  The problem is a race that happens as part of the full-screen process where the swapchain gets created 3200x1672 but, by the time `vkQueuePresent` gets called, the window is 3200x1800.  Sadly, such races are all too common with full-screen.  For some reason, runing it in GDB causes just enough of a delay somewhere that the race doesn't happen and it starts up just fine. :frowning:  This time I'm pretty sure it's not my fault. :stuck_out_tongue_winking_eye:

**Comment by hrydgard** [comment]
Okay, I'll make sure I'm handling the surface query correctly and retry with a different device if it's not there. I'll file an issue for that.

But that race should cause a SUBOPTIMAL when I next present since it'll be the wrong size, and then it should correct itself? Either way, seems avoidable - wanna open a new issue with the details above or should I?

**Comment by jekstrand** [comment]
Done #13060
